src/librustc_mir/interpret/intern.rs

```diff
@@ -0,0 +1,258 @@
+                }
+                // treat dangling pointers like other statics
+                // just to stop trying to recurse into them
+                return Ok(Some(IsStaticOrFn));
+            },
+        };
+        // This match is just a canary for future changes to `MemoryKind`, which most likely need
+        // changes in this function.
+        match kind {
+            MemoryKind::Stack | MemoryKind::Vtable => {},
+        }
+        // Ensure llvm knows to only put this into immutable memory if the value is immutable either
+        // by being behind a reference or by being part of a static or const without interior
+        // mutability
+        alloc.mutability = mutability;
+        // link the alloc id to the actual allocation
+        let alloc = tcx.intern_const_alloc(alloc);
+        self.leftover_relocations.extend(alloc.relocations.iter().map(|&(_, ((), reloc))| reloc));
+        tcx.alloc_map.lock().set_alloc_id_memory(ptr.alloc_id, alloc);
+        Ok(None)
+    }
+}
+
+impl<'rt, 'a, 'mir, 'tcx>
+    ValueVisitor<'a, 'mir, 'tcx, CompileTimeInterpreter<'a, 'mir, 'tcx>>
+for
+    InternVisitor<'rt, 'a, 'mir, 'tcx>
+{
+    type V = MPlaceTy<'tcx>;
+
+    #[inline(always)]
+    fn ecx(&self) -> &CompileTimeEvalContext<'a, 'mir, 'tcx> {
+        &self.ecx
+    }
+
+    fn visit_aggregate(
+        &mut self,
+        mplace: MPlaceTy<'tcx>,
+        fields: impl Iterator<Item=EvalResult<'tcx, Self::V>>,
+    ) -> EvalResult<'tcx> {
+        if let Some(def) = mplace.layout.ty.ty_adt_def() {
+            if Some(def.did) == self.ecx.tcx.lang_items().unsafe_cell_type() {
+                // we are crossing over an `UnsafeCell`, const qualif ensures that this can only
+                // exist in the base constant, not behind references. Double checking here to make
+                // sure everything is alright
+                assert_eq!(self.mutability, Mutability::Mutable);
+            }
+        }
+        self.walk_aggregate(mplace, fields)
+    }
+
+    fn visit_primitive(&mut self, mplace: MPlaceTy<'tcx>) -> EvalResult<'tcx> {
+        // Handle Reference types, as these are the only relocations supported by const eval.
+        // Raw pointers (and boxes) are handled by the `leftover_relocations` logic.
+        let ty = mplace.layout.ty;
+        if let ty::Ref(_, _, mutability) = ty.sty {
+            match (self.mode, mutability) {
+                // all is "good and well" in the unsoundness of `static mut`
+                (InternMode::StaticMut, _) => {},
+                // immutable references are fine everywhere
+                (InternMode::StaticOrConst, hir::Mutability::MutImmutable) => {},
+                // we statically prevent `&mut T` via `const_qualif`
+                (InternMode::StaticOrConst, hir::Mutability::MutMutable) =>
+                    bug!("const qualif failed to prevent mutable references"),
+            }
```

Comment:
Oh, so that extra thing is just for ICEing here?  Seems like a waste TBH.

"mode" is not a good name then though I think, that's what I would call the visitor's mutability. Maybe "context"? Or something with "item", because it just tracks which kind of item we started out with when doing the interning.

But really I feel the kind of item shouldn't matter. Well maybe except that `const` should be recursively immutable. But the rules for `const` and `static` are different, only the latter can be an `UnsafeCell`.

Reply:
Okay so I take back the part with the waste, this seems like a good place to enforce that `const` are recursively immutable. I'd move this down to the recursive call (where we compute the `mutability` of the nested interning we are about to kick off) -- this really checks that our recursive interning makes sense more than it checks anything about references themselves.

That would also be good place to explain that we do not have to check "frozen" because const qualif excluded all the cases where that could be a problem, and that the `UnsafeCell` check in `visit_aggregate` double-checks that.